### DISCUSSION TITLE
Service call interceptor/proxy/middleware/hook

### DISCUSSION THREAD

**Discussion by protyposis** [home-assistant/architecture#956] [General]
I'd like to discuss the possibility of allowing integrations to modify (foreign) service calls before they are executed.

Currently, the service call API is very simple and works like this:

 - Integration registers a service through [`hass.services.async_register(domain, service, service_func, schema)` ](https://github.com/home-assistant/core/blob/80a5e341b52d863396c63a69a68d30134f36e8ab/homeassistant/core.py#L1785)
 - Any other actor (e.g., user, integration) calls the service through [`hass.services.async_call(domain, service, service_data, ...)`](https://github.com/home-assistant/core/blob/80a5e341b52d863396c63a69a68d30134f36e8ab/homeassistant/core.py#L1871), and HA core validates the call, fires a service call event, and calls the `service_func`
 - Service calls and resulting state changes can be monitored with `EVENT_CALL_SERVICE` and `EVENT_STATE_CHANGED` listeners

A limitation of this design is that integrations, due to the (async) nature of state changes and events, can only react to changes that happened in the past.

A concrete example is [Adaptive Lighting](https://github.com/basnijholt/adaptive-lighting), which adjusts light brightness and temperature according to the time of day. For example, if a light's last brightness is 100% and a user turns it on at night, it first ramps up to 100% before Adaptive Lighting receives a state change event and issues another service call to adjust it down to 1%. This results in a suboptimal user experience and an inflation of calls and events. Ideally, the light would directly turn on at the desired brightness.

We worked around this by implementing a workaround/hack in https://github.com/basnijholt/adaptive-lighting/pull/628 that directly injects brightness/color attributes into service calls (simplified pseudocode):

```python
from homeassistant.components.light import preprocess_turn_on_alternatives

# HACK: access protected property
existing_service = hass.services._services[LIGHT_DOMAIN][SERVICE_TURN_ON] 

async def service_func_proxy(call: ServiceCall) -> None:
    # HACK: Convert read-only data to writeable dictionary for modification
    data = dict(call.data)

    # Process data
    data[CONF_PARAMS].update({ATTR_BRIGHTNESS: 1, ...})

    # Preprocess again
    preprocess_turn_on_alternatives(hass, updated_service_data)

    # Convert data back to read-only
    call.data = ReadOnlyDict(data)

    # Call original service handler with processed data
    await existing_service.job.target(call)

# Replace original service handler
hass.services.async_register(
    LIGHT_DOMAIN, SERVICE_TURN_ON, service_func_proxy, existing_service.schema
)
```

This works very well, avoids complicated workarounds (e.g., using a separate service which is difficult to integrate into Lovelace etc.), and received positive user feedback. However, it suffers from a few disadvantages:

- Accessing protected `hass` internals could break with a future update.
- Service call events are fired before service is called, thus also before the interceptor applies changes, and contain only the original data. Not sure yet if this is desirable.
- Service call data is processed by HA (filtered, modified, restructured) before the interceptor is called, thus the original data cannot be accessed, and updated data needs reprocessing.
- If the service is re-registered by any integration, the interceptor gets lost. This can be detected, and the interceptor reinstalled, but multiple integrations using this approach could compete with each other.

A properly implemented core API that allows to register some kind of handlers could fix all of this. Do you think this could be a valuable core HA feature? Did I miss how this can already be achieved?

Edit: In the simplest case, the API could be something like `hass.services.async_register_hook(domain, service, service_hook_func)` where `service_hook_func` receives and returns the `service_data`.

**Comment by ScottG489** [comment]
I really like this idea and it's been great for the user experience in Adaptive Lighting.

I think it's also fairly advanced and could lead to a lot of pitfalls. I wonder if a version of this could be implemented as a standalone integration? I think at the very least it could act as a POC and would give a better idea of how this might work out in practice.

My first thought is something to simply override a service with an action. So you could make a service call that took the name of a service and the action (such as a script) to override it with. Does this sound like something doable?

Also I understand allowing the user to do this remapping is a little different from the original idea which was to do this at the API level and not have it be something the user would control directly, but I thought this might be a good starting point. Another difference is this would be a permanent change to the service call, and although I don't fully follow the implementation after reviewing the code, I believe you're just temporarily changing the service call as opposed to this idea which is a permanent rewrite.

Anyways, just a rough idea I thought I'd float by you to hopefully get more traction on this :)

**Reply by protyposis** [reply]
Thank you, I'm not sure if I understand what you are asking, though. Are you proposing an integration which exposes a service that, when called, temporarily replaces the action of any other service with a custom action (e.g., "the next time `light.turn_on` is called, execute `script.my_custom_script` instead")?

**Reply by ScottG489** [reply]
Yes, I think that's roughly what I'm suggesting. Though maybe a bit more so "permanently" than "temporarily".

**Reply by protyposis** [reply]
Ok, that's quite easily possible, but it's a different use-case. What I'm proposing here is a way to modify the _data_ of a service call, in contrast to modifying the service, without having control over the service caller or the called service (e.g., adding a brightness parameter to a turn-on command). Basically a kind of man-in-the-middle "attack", but sanctioned by an official API.

Since you mentioned a POC, I see Adaptive Lighting as the POC. It's implemented there, it's in production use, and we know it works great. We just need an official way to implement it for long-term stability and to fix the disadvantages listed above.
